File: src/main/scala/robosky/ether/world/feature/Clump.scala
package robosky.ether.world.feature

import java.util.Random

import com.google.common.primitives.Floats
import net.minecraft.util.math.{BlockPos, Vec3i}

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

object Clump {
  def of(r: Float): Clump = {
    val ir = Math.ceil(r).toInt
    val result = new Clump((for {z <- -ir to ir
                                 x <- -ir to ir
                                 y <- -ir to ir
                                 d = Math.sqrt(x * x + y * y + z * z).toFloat
                                 if d <= r
                                 } yield new Clump.Entry(x, y, z, d)).toBuffer)

    if (result.entries.isEmpty) result.entries += new Clump.Entry(0, 0, 0, 0)
    result.entries.sortWith((a: Clump.Entry, b: Clump.Entry) => Floats.compare(a.d, b.d) < 0)
    result
  }

  private[feature] class Entry(var x: Int, var y: Int, var z: Int, var d: Float) {
    def pos = new Vec3i(x, y, z)

    def blockPos(x: Int, y: Int, z: Int) = new BlockPos(this.x + x, this.y + y, this.z + z)
  }

}

class Clump(val entries: mutable.Buffer[Clump.Entry] = ArrayBuffer.empty) {
  def removeUniform(rand: Random, dx: Int, dy: Int, dz: Int): BlockPos = {
    val index = rand.nextInt(entries.size)
    entries.remove(index).blockPos(dx, dy, dz)
  }

  def removeGaussian(rand: Random, dx: Int, dy: Int, dz: Int): BlockPos = {
    var index = Math.abs(rand.nextGaussian * entries.size).asInstanceOf[Int]
    if (index >= entries.size) index = 0 //in unlikely scenarios, gaussian numbers can go way outside the bounds.
    if (index < 0) index = 0
    entries.remove(index).blockPos(dx, dy, dz)
  }

  def copy: Clump = new Clump(this.entries.clone())

  def size: Int = entries.size

  def isEmpty: Boolean = entries.isEmpty
}

File: src/main/java/robosky/uplands/world/feature/Clump.java
package robosky.uplands.world.feature;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import com.google.common.primitives.Floats;

import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;

public final class Clump {

    private final List<Entry> entries;

    private Clump(List<Entry> entries) {
        this.entries = entries;
    }

    public static Clump of(float r) {
        int ir = (int)Math.ceil(r);
        List<Entry> entries = new ArrayList<>();
        for(int z = -ir; z <= ir; z++) {
            for(int x = -ir; x <= ir; x++) {
                for(int y = -ir; y <= ir; y++) {
                    float d = (float)Math.sqrt(x * x + y * y + z * z);
                    if(d <= r) {
                        entries.add(new Clump.Entry(x, y, z, d));
                    }
                }
            }
        }
        if(entries.isEmpty()) {
            entries.add(new Clump.Entry(0, 0, 0, 0));
        }
        entries.sort((a, b) -> Floats.compare(a.d, b.d));
        return new Clump(entries);
    }

    public BlockPos removeUniform(Random rand, int dx, int dy, int dz) {
        int index = rand.nextInt(entries.size());
        return entries.remove(index).blockPos(dx, dy, dz);
    }

    public BlockPos removeGaussian(Random rand, int dx, int dy, int dz) {
        int index = (int)Math.abs(rand.nextGaussian() * entries.size());
        if(index >= entries.size()) {
            index = 0; //in unlikely scenarios, gaussian numbers can go way outside the bounds.
        } else if(index < 0) {
            index = 0;
        }
        return entries.remove(index).blockPos(dx, dy, dz);
    }

    public Clump copy() {
        return new Clump(new ArrayList<>(this.entries));
    }

    public int size() {
        return entries.size();
    }

    public boolean isEmpty() {
        return entries.isEmpty();
    }

    private static class Entry {

        private final int x;
        private final int y;
        private final int z;
        private final float d;

        public Entry(int x, int y, int z, float d) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.d = d;
        }

        public Vec3i pos() {
            return new Vec3i(x, y, z);
        }

        public BlockPos blockPos(int x, int y, int z) {
            return new BlockPos(this.x + x, this.y + y, this.z + z);
        }
    }
}
